controller/test/pid_lib/pid_test.cpp

```diff
@@ -0,0 +1,393 @@
+  input = input - 5;
+  double derivative =
+      ((setpoint - input) - (previous_setpoint - previous_input)) /
+      sample_time * 1000.0;
+
+  Hal.delay(sample_time);
+  myPID.Compute();
+
+  EXPECT_NEAR(output, derivative * Kd, OUTPUT_TOLERANCE);
+
+  previous_setpoint = setpoint;
+  previous_input = input;
+  // output being in [0 255], create a positive setpoint derivative in order to
+  // have a positive error derivative, and therefore a positive output
+  setpoint = setpoint + 5;
+  derivative = ((setpoint - input) - (previous_setpoint - previous_input)) /
+               sample_time * 1000.0;
+  Hal.delay(sample_time);
+  myPID.Compute();
+  EXPECT_NEAR(output, derivative * Kd, OUTPUT_TOLERANCE);
+}
+
+TEST(pidTest, CallFasterThanSample) {
+  // This test calls PID::Compute faster than sample time to check this has no
+  // effect on output
+  float Kp = 5.5;
+  float Ki = 1.1;
+  float Kd = 1.5;
+  double output = 0;
+  PID myPID(&input, &output, &setpoint, Kp, Ki, Kd, DIRECT);
+  myPID.SetSampleTime(sample_time);
+  myPID.SetMode(AUTOMATIC);
+  myPID.Compute();
+
+  double last_output = output;
+  // check that call with no time change has no effect
+  myPID.Compute();
+  EXPECT_EQ(output, last_output);
+
+  Time last_time = Hal.now();
+
+  // chose dt to not be a divisor of SampleTime, for better coverage
+  double dt = 6;
+  // Run PID a few times with fast and check output doesn't change unless a
+  // sample time has passed
+  for (int i = 0; i < 100; i++) {
+    Hal.delay(dt);
+    myPID.Compute();
+    if (Hal.now() >= last_time + milliseconds(sample_time)) {
+      last_output = output;
+      last_time = last_time + milliseconds(sample_time);
+    } else {
+      EXPECT_EQ(output, last_output);
+    }
+  }
+}
+
+TEST(pidTest, TaskJitter) {
+  // This test uses integral to check the effect of time between calls on the
+  // PID output. Introducing jitter in call frequency and checking that the
+  // integral takes this jitter into account.
+  float Kp = 0;
+  float Ki = 0.5;
+  float Kd = 0;
+  double output = 0;
+  PID myPID(&input, &output, &setpoint, Kp, Ki, Kd, DIRECT);
+  myPID.SetSampleTime(sample_time);
+  myPID.SetMode(AUTOMATIC);
+  myPID.Compute();
+  // First task, with only Ki set, output = error*sample_time
+  double integral = (setpoint - input) * (sample_time / 1000.0);
+  EXPECT_NEAR(output, integral * Ki, OUTPUT_TOLERANCE);
+  // Advance time with jitter
+  uint32_t dt = sample_time + max_task_jitter;
+  Hal.delay(dt);
+  myPID.Compute();
+  // Expect output to take jitter into account in integral
+  integral += (setpoint - input) * (dt / 1000.0);
+  EXPECT_NEAR(output, integral * Ki, OUTPUT_TOLERANCE);
+  // Advance time and compensate previous jitter to have total time
+  // elapsed = 2*sample time
+  dt = sample_time - max_task_jitter;
+  integral += (setpoint - input) * (dt / 1000.0);
+  Hal.delay(dt);
+  myPID.Compute();
+  EXPECT_NEAR(output, integral * Ki, OUTPUT_TOLERANCE);
```

Comment:
Hm in this case shouldn't the PID just ignore this Compute() because it's not the next sample time yet? I'm again confused how to reconcile this vs. CallFasterThanSample

Reply:
Again, in an ideal world, PID executes at the exact sample time, and I think in the practical world, it is better not to accumulate jitter but rather have only a small jitter on each one.
I admit this is a personal interpretation.

Reply:
I linked an illustration of this in the description for the PR